### ISSUE TITLE
Undesired tile boundaries while displaying MVT layer

### ISSUE THREAD

**Issue by mshubin** [openlayers/openlayers#10348]
**Describe the bug**
ol6 renders undesired tile boundaries while displaying `VectorTileLayer` with `MVT` source using ol-mapbox-style `stylefunction`.

![tile_boundaries](https://user-images.githubusercontent.com/2290300/69809429-07da7c00-11ea-11ea-8273-45879e12205b.png)

**To Reproduce**
1. Go to https://codesandbox.io/s/infallible-sunset-62e1p?fontsize=14&hidenavigation=1&theme=dark
2. Try to zoom-in and zoom-out multiple times. After some zooming manipulations the tile boundaries appear as on screenshot

**Expected behavior**
No tile boundaries should be visible

**PS**
Thank you very much for your awesome work, guys :+1:

**Comment by mike-000** [comment]
I've seen similar and it is the background color showing because polygons end at the edge of tiles instead of being buffered.  If you set a highly contrasting background it will be much more apparent.  It is probably more conspicuous at fractional zoom levels but it could be prevented by better tile design 
https://codesandbox.io/s/mapbox-vector-tiles-gfyfn
![image](https://user-images.githubusercontent.com/49240900/69821501-c05ceb80-11fb-11ea-8cb4-8c539f13448b.png)

**Comment by mshubin** [comment]
Ok, thank you for explanation.
By constraining the fractional zoom with `constrainResolution: true` option it completely resolves the problem :+1: 

It's ok for now but just curious, what do you mean by 'better tile design' ?
I'm still able to reproduce the issue in codesandbox that you provided.

**Comment by mike-000** [comment]
I have seen tiles with polygons for features such as sea divided at tile boundaries which would defeat the benefits of buffering at tile edges.   I have checked my example and the divisions don't match the effect on the styled map, so I am seeing the same fractional zoom effect as you.

![image](https://user-images.githubusercontent.com/49240900/69900358-fb7d2d00-1369-11ea-90e9-47f8f6e753e9.png)

**Comment by jokd** [comment]
It is also apparent in the mapbox-vector-tiles-advanced (https://openlayers.org/en/v6.1.1/examples/mapbox-vector-tiles-advanced.html) example:
![image](https://user-images.githubusercontent.com/6848075/70052891-8feac800-15d4-11ea-9ee4-86789829ee1a.png)
Windows 10, Chrome 73

**Comment by ahocevar** [comment]
Make sure that no browser zoom is active. You should be able to reset any browser zoom by pressing Ctrl+0 (or Command+0 on a Mac).

**Comment by jokd** [comment]
Unfortunately I was at 100% already.

**Comment by ahocevar** [comment]
@jokd Are you seeing the same with the https://openlayers.org/en/v6.1.1/examples/mapbox-vector-tiles.html example, or is that working as expected?

Note: I am unable to reproduce the issue with Chrome 78, and I am also not seeing it on any other browser.

**Comment by jokd** [comment]
@ahocevar That works as expected. A collegue of mine experience the same issue and I can also reproduce it in Firefox 70

**Comment by ahocevar** [comment]
@jokd If that other example works, then what @mike-000 says is the case. Using only a subset of the available tiles, at lower resolutions than they are intended for (which is what the mapbox-vector-tiles-advanced example does), can lead to such rendering problems.

**Comment by jokd** [comment]
I can see the issue on maps without custom tileUrlFunction but I suppose it is a serverside issue then. Would it be possible to write a tileUrlFunction for tileservers that produces these kind of errors?

**Comment by jokd** [comment]
Came back to this and setting declutter: true solved the issue for me.

**Comment by Akoneto9** [comment]
Nice

**Comment by sookoll** [comment]
I also see tile edges when using pattern as fill color.
![image_2020_03_02T12_51_33_585Z](https://user-images.githubusercontent.com/1793122/75680397-33ae4780-5c9a-11ea-80cb-c19ad33d40ab.png)

I calculate pattern like that:
```
function createPattern (conf) {
  const patternCanvas = document.createElement('canvas')
  const pctx = patternCanvas.getContext('2d', { antialias: true })
  const division = 2
  const lineWidth = conf.step[0] || 1
  const offset = conf.step[1] || 2
  const size = offset * division
  patternCanvas.width = size
  patternCanvas.height = size
  pctx.strokeStyle = conf.color || '#fff'
  pctx.lineWidth = lineWidth
  pctx.beginPath()
  pctx.moveTo(offset, 0)
  pctx.lineTo(0, offset)
  pctx.moveTo(offset * 2, 0)
  pctx.lineTo(0, offset * 2)
  pctx.moveTo(offset * 2, offset)
  pctx.lineTo(offset, offset * 2)
  pctx.stroke()
  return pctx.createPattern(patternCanvas, 'repeat')
}
```
Also, is it possible to label vector tile with one label per polygon and not repeated on each tile?

**Comment by ahocevar** [comment]
@sookoll Make sure your pattern's width and height are a factor of two. See https://openlayers.org/en/latest/apidoc/module-ol_colorlike.html#~ColorLike.

**Comment by msch-alpgis** [comment]
I'm still able to reproduce the problem with both sandboxes posted in previous comments:
 * https://codesandbox.io/s/mapbox-vector-tiles-gfyfn
 * https://codesandbox.io/s/infallible-sunset-62e1p?fontsize=14&hidenavigation=1&theme=dark

Tested with ol 6.1 and 6.3 in Chrome 81 and FF 76.

![image](https://user-images.githubusercontent.com/5549925/81171421-a5f21f80-8f9c-11ea-81f8-644a76ab5270.png)

**Comment by ahocevar** [comment]
@msch-alpgis As mentioned previously, this is a problem with the tiles. Polygons need to have a buffer. Are you seeing the issue also in https://openlayers.org/en/latest/examples/mapbox-vector-tiles.html?

**Comment by msch-alpgis** [comment]
No, the example seems to be ok. But this is in 4326 too. Take a look at https://www.maptiler.com/coordinate-systems/. There you'll find a mapbox map and an openlayers map with the same vector tiles as source. The maps are both in EPSG:2056. But only in the openlayers map i can see the artifacts. Best you zoom into a lake to see the lines.

**Comment by ahocevar** [comment]
The example I linked is in EPSG:3857. This one is in EPSG:4326: https://openlayers.org/en/latest/examples/vector-tiles-4326.html. Do you see borders on that one? I cannot comment on what you see on www.maptiler.com, because I do not know if they are using the latest version of OpenLayers.

**Comment by msch-alpgis** [comment]
This example works as well. I put a simple example together, where you should see the behaviour: https://codesandbox.io/s/amazing-cray-87cdr?file=/src/index.js

When i zoom in and out a little i get something similar to:
![image](https://user-images.githubusercontent.com/5549925/81177116-734d2480-8fa6-11ea-8af4-0b15b762d972.png)

**Comment by ahocevar** [comment]
Looks like the tiles used there do not have a buffer. I also get cut-off labels there, which is a clear indicator for that.

**Comment by msch-alpgis** [comment]
Okay, thank you so far! I will write an update if i have any news on this. My problem is still:
 * vector tiles from geoserver in EPSG:4326 works fine
 * vector tiles from geoserver in EPSG:2056 have the mentioned lines

**Comment by msch-alpgis** [comment]
Giving up on this. I just note what I did and what results I got for others landing on this page.

Imported some OSM data into postgis with osm2pgsql. To test I imported the data in four different projections into separated databases:

* EPSG:3857
* EPSG:4326
* EPSG:2056
* EPSG:21781

Then I created a simple view and from this vector tiles with [GeoServer](https://geoserver.org/) and [T-Rex Tileserver](https://t-rex.tileserver.ch). From both servers the tiles in EPSG:3857 and EPSG:4326 are working as expected. But not for the other two projections. There i get from both servers (or openlayers) the visible "tileboundaries" (It's already visible in GeoServers preview if your eyes are still good). If I check the tiles as GeoJSON in QGIS, then i can see that there is a buffer.

I think it's something with the TileGrid/GridSet but I'm not able to find the solution.
